### PR TITLE
Improve performance of `brew info` by caching compiled regexes

### PR DESCRIPTION
**Pull request by ThatsJustCheesy**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

I've taken an interest in speeding up large `brew info` dumps (for my GUI client again).

Profiling with `time brew prof --stackprof -- info --json=v2 --eval-all`, I found:

- `Cask#os_versions` took up a large portion of the total time
   - Cask processing took 49% overall when the [`"versions"` key was evaluated](https://github.com/Homebrew/brew/blob/560f5717ad4dc1e3feed9f3c0b7c84aa1498b84c/Library/Homebrew/cask/cask.rb#L239), compared to 12% when skipped (by commenting it out)
   -  `CaskLoader.for` took up [about 45% of the cask processing time](https://user-images.githubusercontent.com/16456186/198201309-1d5e9646-f0b9-416e-b661-2b80a4ef1a01.png)
       - This is time spent determining how to load the cask, not actually loading it
       - `FromURILoader.can_load?` and `FromContentLoader.can_load?` took a [suspiciously large portion of that time](https://user-images.githubusercontent.com/16456186/198201950-eca40932-3b32-4238-94ff-4d3241151b25.png)

Those last two methods have one thing in common: compiling regexes.

`master` branch:

```
$ time brew prof --stackprof -- info --json=v2 --eval-all

# …
________________________________________________________
Executed in   28.51 secs    fish           external
   usr time   21.10 secs    0.18 millis   21.10 secs
   sys time    2.80 secs    1.18 millis    2.79 secs
```

Caching compiled regexes in those two methods:

```
$ time brew prof --stackprof -- info --json=v2 --eval-all

# …
________________________________________________________
Executed in   22.92 secs    fish           external
   usr time   16.27 secs  191.00 micros   16.27 secs
   sys time    2.58 secs  867.00 micros    2.58 secs
```

and `CaskLoader.for` now takes [about 16% of the cask processing time](https://user-images.githubusercontent.com/16456186/198202710-9e59886e-d1d0-4fdc-b71b-0f969e06c642.png), rather then 45%.

---

tl;dr this PR changes almost nothing, but improves performance of `brew info --json=v2 --eval-all` by about 15–20%

(with apologies for all the %s 😄)

### CODE REVIEW

**Review by MikeMcQuaid — APPROVED** [review]
Very nice! Will happily accept more performance PRs like this that benefit your client.
